# Dzień piąty DWthon 👋 - Hack Outside the Box 

### Celem jest poznanie, czym jest tak zwany "aha moment"!

Czy wiesz co to jest "aha-moment"? To jest taki moment, kiedy mówisz... "aha" 😁. Czyli coś Cię olśniło i rzeczywistość (a przynajmniej jej kawałek) stał się bardziej dla Ciebie zrozumiały.

O "aha-momencie" można długo mówić, lepiej podam Ci kilka przykładów.


### Facebook (7 przyjaciół w 10 dni)
Jeśli użytkownik dodaje co najmniej **10 znajomych** w ciągu **pierwszych 7 dni** od rejestracji, wtedy ten użytkownik będzie bardziej zaangażowany i na dłużej zostanie w Facebook.

### Dropbox (1 plik, 1 folder, 1 urządzenie)
Jeśli użytkownik doda co najmniej jeden plik, jeden folder i zsynchronizuje się z co najmniej jednym urządzeniem, to wtedy ten użytkownik zostanie na dłużej.

### LinkedIn - X połączeń w Y dni


To są proste reguły, które potrafią opisać najważniejsze, co ma się wydarzyć w firmie, aby firma rosła. Innymi słowy to jest stosowanie wprost reguły 80/20.


Trudność polega na tym, że aby znaleźć tę prostą regułę, to... no właśnie, jak można ją znaleźć? Przecież takich prostych reguł może być tysiące lub nawet miliony, zabraknie nam czasu, aby je wszystkie sprawdzić. Tutaj na pomoc nam może przyjść `Machine Learning`.

### Zanim zaczniemy | ważne ❗

Ankieta numer 5! Każdy dzień DWthon - hack outside the box ma dedykowaną ankietę. Wypełnij ją zanim przejdziesz do nauki. Zajmie Ci to tylko 3 minutki ;) 

### [Wypełnij ankietę](https://bit.ly/3qxZd9S)  i mierz deltę swego rozwoju :) 

### Gdzie zadawać pytania ❓ 

Jeśli napotkasz trudności podczas wykonywania zadań z tego notebooka, to koniecznie napisz w kanale [dwthon_day5](https://bit.ly/3qEBEMv)

Pamiętaj, aby szczegółowo doprecyzować, z czym masz problem. Najlepiej wrzuć screen z kodem swoim lub błędem, który widzisz i napisz, którego zadania dotyczy :)

In [23]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/EeFJLSS3_x0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

### Krok po kroku 

Nagrałem dla Ciebie także materiał wideo "krok po kroku". W wideo poniżej znajduje się dokładnie to, co w tym notebooku tylko tłumaczę wszystko, aby ułatwić Ci pracę i zrozumienie zagadnień i zadań :) Obejrzyj, jeśli potrzebujesz takiego dodatkowego wsparcia. 

In [24]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/7UTzlVkK17k" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## Biblioteki
Potrzebujemy `pandas` i jeszcze trochę więcej innych bibliotek. 
* `DecisionTreeClassifier` - do trenowania modelu
* `sklearn.model_selection` walidacja modelu
* `gc` do sprzątania w pamięci RAM

Dodatkowo użyjemy nowego bardziej zaawansowanego modelu `xgboost` oraz biblioteki `eli5` do podglądania ważności cech.
`from collections import Counter` przyda nam się do wygodnego zliczenia liczby wystąpień (np. ile razy klient kupił produkt x).

In [1]:
import pandas as pd
import numpy as np

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
import xgboost as xgb
import eli5 
from collections import Counter

import gc

## Wczytajmy dane

In [2]:
df = pd.read_hdf("../input/data.h5")

print(df.shape)
df.sample(5)

(820906, 9)


order_id  customer_id  product_id  quantity  price_unit  price_total  \
646414     42689         4622         639         1         295          295   
593006     38874         5298        1021         2         749         1498   
57292       4115          807        1335        12         165         1980   
295509     19593         1211         123        18         254         4572   
22785       1603          844        2268        24          42         1008   

        country_id          order_date  is_canceled  
646414           0 2010-08-17 13:35:00        False  
593006           0 2010-06-17 16:20:00        False  
57292            0 2011-02-14 15:35:00        False  
295509           0 2011-10-16 13:23:00        False  
22785            0 2010-12-16 13:44:00        False

## Przygotowujemy dane


To jest kod z poprzedniej lekcji. Jedynie poukładamy kod w funkcji, aby łatwiej było nam eksperymentować później.

In [3]:
df_customers = (
    df[ ["price_total", "customer_id"] ]
    .groupby("customer_id")
    .agg("sum")
    .reset_index()
    .sort_values(by="price_total", ascending=False)
    .rename(columns={"price_total": "customer_price_total"})
)


df_customers["cumsum"] = df_customers["customer_price_total"].cumsum()
value_80prc = int(df["price_total"].sum() * 0.8)
df_customers["most_revenue_customer"] = df_customers["cumsum"] < value_80prc


top_customers = set(df_customers[ df_customers["most_revenue_customer"] ]["customer_id"].unique())

del df_customers
gc.collect()

def feature_engineering(df):
    df_customers = (
        df
        .groupby("customer_id", as_index=False)
        .agg(
            count_orders=("order_id", lambda x: len(set(x))),
            count_unq_products=("product_id", lambda x: len(set(x))),
            sum_quantity=("quantity", np.sum),
            sum_price_unit=("price_unit", np.sum),
            sum_price_total=("price_total", np.sum),
            count_unq_countries=("country_id", lambda x: len(set(x))),
            prob_canceled=("is_canceled", np.mean)
        )
    )
    
    
    
    return df_customers


def get_feats(df_customers, black_list=["most_revenue_customer"]):
    feats = list(df_customers.select_dtypes([np.number, bool]).columns)
    return [x for x in feats if x not in black_list]

def get_X_y(df_customers, top_customers, feats):
    df_customers["most_revenue_customer"] = df_customers["customer_id"].map(lambda x: x in top_customers)
    
    X = df_customers[feats].values
    y = df_customers["most_revenue_customer"].values
    
    return X, y


def train_and_get_scores(model, X, y, scoring="accuracy", cv=5):

    scores = cross_val_score(model, X, y, scoring=scoring, cv=cv)
    return np.mean(scores), np.std(scores)

## Model 1

☝️ Dokładnie to, co zrobiliśmy ostatnio.

In [4]:
df_customers = feature_engineering(df)
feats = get_feats(df_customers)
X, y = get_X_y(df_customers, top_customers, feats)
model = DecisionTreeClassifier(max_depth=5)

train_and_get_scores(model, X, y)

(0.9996598639455783, 0.0004165798882284457)

Mamy wynik 99.9%.

Można się cieszyć, ale... tak naprawdę opowiem Ci o jednym mechanizmie, który regularnie stosuję. Nazywam go wahadło.

Idea polega na tym, aby najpierw być dużym optymistą i bardzo upraszczać rzeczywistość na początku. Następnie, jak już mamy dobry wynik, to wchodzimy w "zupełnie inne buty". Wchodzimy w rolę aktywnego "krytyka", który próbuje zrobić wszystko, aby udowodnić, że to, co zrobiliśmy nie ma sensu. Powtarzamy tak w kółko. Robimy kilka iteracji :)  

Najważniejsze jest to, że dzięki przełączaniu się pomiędzy rolami, można bardzo szybko biec od przodu, ale z drugiej strony mieć poczucie, że to nadal ma sens!

## 🧠 Włączmy myślenie krytycznie

Na ten moment wykonaliśmy pierwszy krok, czyli optymistyczny i dostaliśmy wynik 99%. Bardzo dobrze, ale... Problem polega na tym, że `price_unit_total`jest zbyt mocną cechą. Czyli ta cecha wprost mówi, czy klient należy do segmentu `top_customers` czy nie. Dlatego teraz musimy co najmniej tę cechę wyłączyć i sprawdzić, jak zachowa się model.


Wystarczy dodać `sum_price_total` do listy, którą będziemy ignorować. 

Zobacz 👀.


## Model 2

In [5]:
df_customers = feature_engineering(df)
feats = get_feats(df_customers, black_list=["most_revenue_customer", "sum_price_total"])
X, y = get_X_y(df_customers, top_customers, feats)
model = DecisionTreeClassifier(max_depth=5)

train_and_get_scores(model, X, y)

(0.7725821392386741, 0.2654553956998111)

Jak widzisz, jakość modelu nam spadła do 77%. Z jednej strony smutna sprawa 😂, ale z drugiej strony używamy modelu ML, aby znajdował mniej oczywiste zależności. Bo te oczywiste człowiek sam łatwo znajdzie sam 🤦‍♂️.


### Myślenie optymistyczne
To teraz wróćmy znów do bycia optymistą. Pamiętasz na czym polega zasada wahadła? Wahamy się tam i z powrotem ;) 


Można zmienić model na bardziej zaawansowany i to powinno poprawić wynik. Użyjmy `xgboost`. Ta nazwa może Ci nic nie mówić i to jest ok na tym etapie 👌.

## Model 3

Zwróć uwagę, że zmieniła się ta linijka: `model = xgb.XGBClassifier(max_depth=5, n_estimators=50, learning_rate=0.3)`.

In [6]:
df_customers = feature_engineering(df)
feats = get_feats(df_customers, black_list=["most_revenue_customer", "sum_price_total"])
X, y = get_X_y(df_customers, top_customers, feats)
model = xgb.XGBClassifier(max_depth=5, n_estimators=50, learning_rate=0.3)

train_and_get_scores(model, X, y)

/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)
/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[22:27:54] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[22:27:55] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)
/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[22:27:55] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[22:27:55] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[22:27:55] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


(0.8106769431176726, 0.08696513375768444)

## Mamy wynik 81%.

Spróbujmy zrobić pierwszą interpretację modelu i zobaczyć, co wg nieg0 jest ważne.

In [8]:
model.fit(X, y)

/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[22:28:38] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.3, max_delta_step=0, max_depth=5,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=50, n_jobs=4, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [9]:
eli5.show_weights(model, feature_names=feats)

### Wstępnie wnioski:
#### TOP3 
- `sum_quantity` brzmi jako najważniejsza cecha*
- `count_orders`
- `sum_price_unit`

Wyglada na to, że cecha `count_unq_countries` jest mało przydatna.


☝️ Trzeba uważać na to, co jest "ważne", bo doświadczenie pokazuje, że to może być dość zmienne i w tej zmienności trzeba nauczyć się poruszać.


## Następny krok

Możemy teraz trochę pofilozofować nad tymi kilkoma cechami, ale to jest mało istotne. Te cechy same w sobie są dość oczywiste. Nie ma poczucia "dużej" wartości dodanej. 

Ciekawostką okazało się, że `sum_quantity` jest istotna. Tylko tutaj też nie ma dużego zaskoczenia.


Wygenerujmy zatem więcej cech, niech model znajdzie na co warto zwrócić naszą uwagę :) 


## Produkty


Możemy dodać dla każdego produktu osobną kolumnę i jako wartość w tej kolumnie dodać, ile produktów już było kupionych (czyli suma). W ten sposób chcemy znaleźć, które produkty są "najciekawsze", warte tego, aby zwrócić na nie uwagę.

☝️ Oczywiście możemy sprawdzić każdy produkt manualnie, ale mamy ich ponad 3.8k, może to nam trochę zająć czasu. Użyjmy lepiej do tego ML, będzie szybciej, sprawniej i efektywniej.

Tylko najpierw trzeba przygotować dane. Zmieńmy naszą funkcję `feature_engineering`. 

In [10]:
def feature_engineering(df):
    
    def counter(vals):
        cntr = Counter()
        cntr.update(vals)
        return cntr
    
    df_customers = (
        df
        .groupby("customer_id", as_index=False)
        .agg(
            count_orders=("order_id", lambda x: len(set(x))),
            count_unq_products=("product_id", lambda x: len(set(x))),
            count_by_products=("product_id", lambda x:  counter(x) ),
            sum_quantity=("quantity", np.sum),
            sum_price_unit=("price_unit", np.sum),
            sum_price_total=("price_total", np.sum),
            count_unq_countries=("country_id", lambda x: len(set(x))),
            prob_canceled=("is_canceled", np.mean)
        )
    )
    
    
    
    return df_customers


df_customers  = feature_engineering(df)
df_customers.head()

customer_id  count_orders  count_unq_products  \
0            0           159                  93   
1            1            40                 184   
2            2            30                 139   
3            3             9                  40   
4            4            14                  11   

                                   count_by_products  sum_quantity  \
0  {0: 89, 1: 73, 2: 69, 3: 68, 4: 32, 5: 36, 6: ...         22976   
1  {8: 8, 9: 4, 10: 4, 11: 4, 12: 3, 13: 15, 14: ...          3514   
2  {25: 9, 26: 7, 27: 7, 28: 3, 29: 4, 30: 10, 31...          8971   
3  {44: 7, 1781: 1, 1179: 1, 112: 1, 508: 3, 581:...          1108   
4  {54: 14, 1496: 2, 2288: 1, 2315: 1, 3432: 1, 4...           390   

   sum_price_unit  sum_price_total  count_unq_countries  prob_canceled  
0          895827          5746671                    1       0.010966  
1          205962           910113                    1       0.097727  
2           95500          1337464                    1       0.009050  
3           18626           226381                    1       0.000000  
4           20028           346978                    1       0.307692

Jak widzisz pojawiła się nowa kolumna `count_by_products`. W środku jest słownik. Zobaczmy dokładniej 👀.

In [11]:
df_customers["count_by_products"]

0       {0: 89, 1: 73, 2: 69, 3: 68, 4: 32, 5: 36, 6: ...
1       {8: 8, 9: 4, 10: 4, 11: 4, 12: 3, 13: 15, 14: ...
2       {25: 9, 26: 7, 27: 7, 28: 3, 29: 4, 30: 10, 31...
3       {44: 7, 1781: 1, 1179: 1, 112: 1, 508: 3, 581:...
4       {54: 14, 1496: 2, 2288: 1, 2315: 1, 3432: 1, 4...
                              ...                        
5874    {1709: 1, 409: 1, 170: 1, 662: 1, 253: 1, 255:...
5875    {2368: 1, 300: 1, 1385: 1, 2412: 1, 2059: 1, 1...
5876    {1827: 1, 869: 1, 2157: 1, 335: 1, 1931: 1, 21...
5877    {1038: 1, 701: 2, 495: 1, 353: 1, 189: 1, 484:...
5878                                             {694: 2}
Name: count_by_products, Length: 5879, dtype: object

W praktyce to oznacza, że  `customer_id=0` kupił np. `product_id=0` - 89 razy, `product_id=1` - 73 razy, `product_id=3` - 68 razy itd.


Fajnie byłoby nam wypakować słownik do osobnych kolumn, na szczęście da się to zrobić bardzo łatwo. ⌛️ Tylko poczekaj chwilkę na wykonanie.

In [12]:
df_count_products = df_customers["count_by_products"].apply(pd.Series).fillna(-1)
df_count_products.columns = ["product_{}".format(x) for x in df_count_products.columns]

df_count_products.head(5)

product_0  product_1  product_2  product_3  product_4  product_5  \
0       89.0       73.0       69.0       68.0       32.0       36.0   
1        6.0       -1.0        1.0       -1.0       -1.0       -1.0   
2       -1.0       -1.0       -1.0       -1.0       -1.0       -1.0   
3       -1.0       -1.0       -1.0       -1.0       -1.0       -1.0   
4       -1.0       -1.0       -1.0       -1.0       -1.0       -1.0   

   product_6  product_7  product_8  product_9  ...  product_4216  \
0       34.0       35.0       18.0       -1.0  ...          -1.0   
1       -1.0       -1.0        8.0        4.0  ...          -1.0   
2       -1.0       -1.0       -1.0       -1.0  ...          -1.0   
3       -1.0       -1.0       -1.0       -1.0  ...          -1.0   
4       -1.0       -1.0       -1.0       -1.0  ...          -1.0   

   product_4219  product_4220  product_4222  product_4226  product_4231  \
0          -1.0          -1.0          -1.0          -1.0          -1.0   
1          -1.0          -1.0          -1.0          -1.0          -1.0   
2          -1.0          -1.0          -1.0          -1.0          -1.0   
3          -1.0          -1.0          -1.0          -1.0          -1.0   
4          -1.0          -1.0          -1.0          -1.0          -1.0   

   product_4236  product_4237  product_4238  product_4239  
0          -1.0          -1.0          -1.0          -1.0  
1          -1.0          -1.0          -1.0          -1.0  
2          -1.0          -1.0          -1.0          -1.0  
3          -1.0          -1.0          -1.0          -1.0  
4          -1.0          -1.0          -1.0          -1.0  

[5 rows x 3878 columns]

Mamy teraz osobny `dataframe` z 3878 kolumnami. Każda kolumna to jest `product_id`. Wartość -1 oznacza brak informacji, czyli  klient nie kupował tego produktu, natomiast każda inna liczba (większa niż 0) oznacza, ile razy ten produkt był kupiony (to jest trochę optymistyczne zdanie, ale na razie możemy przyjąć, że jest ok).

☝️ Zwróć uwagę, że pojawiła się linijka zmiany nazwy produktu. `df_count_products.columns = ["product_{}".format(x) for x in df_count_products.columns]` W tym miejscu robimy bardzo prostą rzeczy, zamiast samego ID, np. 0 czy 2, dodajemy przedrostek, aby kolumna nazywała się  `product_0` czy `product_2`. To nam przyda się za chwilę.


### Następny krok

Połączmy `df_count_products` z `df_customers` i wytrenujmy model. Aby to połączyć użyjemy konstrukcji `concat()`. Wynikiem będzie to, że dokleimy to "z boku". Innymi słowy liczba wierszy zostanie taka sama, liczba kolumn powiększy się o 3878 😱. Model pewnie ucieszy się, będzie miał, co optymalizować.

Łączymy!

In [13]:
df_customers = pd.concat([df_customers, df_count_products], axis=1)
df_customers.shape

(5879, 3887)

Tak jak spodziewaliśmy się, liczba wierszy została taka sama - 5879 (bo liczba wierszy to są klienci, nowych klientów nie przybyło). Liczba kolumn wzrosła nam do 3887. Możesz o tym pomyśleć, że teraz każdy klient posiada 3887 cech, czyli charakterystyk.

## 🤖 Czas na trenowanie modelu

Najpierw sprawdźmy, czy jakość modelu się nam poprawiła.

Uwaga! Model potrzebuje czasu na trenowanie, więc możesz spokojnie przez chwilę odpocząć. Model wykonuje "prostą" i powtarzalną pracę, natomiast po Twojej stronie jest myślenie koncepcyjnie. Przyzwyczajaj się, to jest przyszłość, która już do nas przyszła :)

In [14]:
#df_customers = feature_engineering(df)
feats = get_feats(df_customers, black_list=["most_revenue_customer", "sum_price_total"])
X, y = get_X_y(df_customers, top_customers, feats)
model = xgb.XGBClassifier(max_depth=5, n_estimators=50, learning_rate=0.3)

train_and_get_scores(model, X, y)

/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[22:38:11] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[22:38:25] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[22:38:40] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[22:38:54] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[22:39:08] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


(0.8474117817339701, 0.08924212099402075)

Jakość modelu poprawiła się, mamy już ~85%. 

Przypomnę, że same drzewa decyzyjne dały nam 77%. XGBoost na podstawowych cechach dał nam 81% i teraz mamy 85%.

Może się wydawać, że to słabo, bo mieliśmy 99,9%. Chcę taki wynik znowu! 

Być może to jest przykra wiadomość, ale często model "stopuje" się właśnie w okolicy 60-70% i potem trzeba naprawdę postarać się, aby wycisnąć więcej. No chyba, że użyjemy oczywistych zależności, ale to jest mało ciekawe. Lepiej jednak pozwolić modelowi znaleźć coś, co może zaskoczyć.

Wytrenujmy model teraz niezależnie, aby można było zajrzeć i sprawdzić, co dla niego jest ważne.

In [15]:
model.fit(X, y)

/opt/conda/lib/python3.8/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[22:40:54] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.3, max_delta_step=0, max_depth=5,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=50, n_jobs=4, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [17]:
eli5.show_weights(model, feature_names=feats, top=50)

No teraz już przynajmniej mamy na co popatrzeć :). Widać, że ważność produktu `product_545` oraz `product_99` jest ponad `count_orders`. Oczywiście, tak jak mówiłem, tutaj trzeba ostrożnie wyciągać wnioski. Natomiast to, co jest jednoznaczne to to, że `product_id=545` zasłużył na to, aby zwrócić na niego uwagę 🥳. To sprawdźmy :).

In [18]:
customer_ids_by_product = set(df[ df["product_id"] == 545 ]["customer_id"].unique())
len(customer_ids_by_product )

168

Ten produkt kupiło nie tak wielu klientów, tylko 168. Sprawdźmy, ilu z nich należy do segmentu `most_revenue_customer`.

In [19]:
df_customers[ df_customers.customer_id.isin(customer_ids_by_product) ]["most_revenue_customer"].mean()

0.4166666666666667

No widzimy, że mamy 42% klientów, którzy kupili co najmniej `product_id=545` i należą do segmentu `most_revenue_customer`.

Przypomnę, że normalnie mamy 22% klientów, którzy należą do segmentu `most_revenue_customer`.

In [20]:
df_customers["most_revenue_customer"].mean()

0.225208368770199

Możemy na szybko postawić hipotezę (do zbadania) -  jeśli klient kupił `product_id=545` to prawdopodobieństwo, że będzie należał do segmentu `most_revenue_customer` prawie podwaja się (od 23% do 42%). Brzmi jako "aha moment". Tę hipotezę warto jeszcze zbadać dokładniej, ale jeżeli ona faktycznie działa, to oznacza, że warto sprawić, aby więcej ludzi kupowało właśnie `product_id=545`.

#### Sprawdźmy, ile kosztuje ten produkt. Może to jakiś luksus ⚜️.

In [21]:
df[ df.product_id == 545 ]["price_unit"].value_counts()

125    126
39     103
42       5
246      3
106      1
Name: price_unit, dtype: int64

Z ceną tego produktu bywa różnie, ale najczęściej cena jednostkowa, to jedynie 1,25 lub 3,9 😱. Taki ciekawy wniosek udało się znaleźć w tak łatwy sposób.

## Zadanie 5.1


Twoim zadaniem jest zrobić podobną analizę dla produktu.


### 💡 Podpowiedzi: 

1. Przygotuj `df_products`
2. Dodaj więcej cech, przed dodaliśmy `product_id`, ale możesz zrobić symetryczne lub wymyśl coś innego (może `order_id`) lub jeszcze coś innego, jest sporo pomysłów :) 
3. Wytrenuj model. 
4. Zobacz, co dla modelu było najbardziej istotne.

In [ ]:
### YOUR CODE HERE

Jak **wykonasz** zadanie, należy zrobić:

1. Zrzut ekranu na którym widać rozwiązania, wtedy dostaniesz bonus :)
2. Wrzucić rozwiązanie na slacku do pokoju **[#dwthon_day5_done](https://bit.ly/3teY3l8)**
3. Dostać bonus za dobrą robotę 💪 i jutro będzie kolejna porcja.

*Uwaga! Jeśli masz problem z jakimś zadaniem, czegoś nie wiesz, to pamiętaj, że możesz zadawać pytania! Na tym polega nauka :) Pytania związane z 2 dniem DWthon zadawaj w kanale [#dwthon_day5](https://bit.ly/3qEBEMv)*.

## 🧠 Włączmy myślenie krytycznie


Ta cała nasza analiza była oparta na `price_total`, czyli przychodach. Co ze zwrotami? Być może nastąpił czas nieco bardziej wejść do króliczej nory?

Na co jeszcze warto zwrócić krytyczną uwagę?  Podpowiem, że jeszcze jest sporo rzeczy ;)

## 😇 Włączmy myślenie kreatywne

Pomyśl, jak możesz wykorzystać tę wiedzę, którą tutaj udało Ci się zdobyć?

Najlepiej, aby padły przykłady z Twojego podwórka. Jeśli nie masz (ale pomyśl, bo to tylko może się wydawać, że nie ma) to pomyśl o swoim otoczeniu.

Od razu powiem, że nie chodzi tylko o e-commerce. To może być dowolna branża. Reguła 80/20 jest uniwersalna. To samo dotyczy "aha-momentów".

## 🤝🗣️ Pomyśl o kooperacji i komunikacji

#### ☝️ Podziel się swoimi przemyśleniami na slacku **[#dwthon_day5_ideas](https://bit.ly/3qFcb5R)**. Jeśli masz pomysł, ale nie masz kodu - poproś na slacku o pomoc, aby ktoś pomógł Ci napisać kod. Natomiast jeśli masz kod, to koniecznie podziel się. Dzięki temu ktoś inny będzie w stanie łatwo powtórzyć i podzielić się swoimi przemyśleniami. Dzięki temu każdy zyskuje :).


In [22]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/qC2MZcjGfcM" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## Twój Feedback i delta rozwoju 

Daj mi znać koniecznie, co sprawiło Ci przyjemność, co trudność oraz czego udało się nauczyć w tej krótkiej [ankiecie](https://bit.ly/3rEnOLD). Pozwoli mi ona ulepszyć zadania, a Tobie lepiej trakować swój rozwój: win-win 😇

## 💪 Pochwal się światu, że DWthon za Tobą 💪 

Udało Ci się! Jesteś na mecie, koniecznie powiedz o tym innym. Być może kogoś to zmotywuje do działania, a dodatkowo warto mówić głośno o nowych umiejętnościach i doświadczeniach, bo nigdy nie wiesz, jaka furtka może się otworzyć ;) 

Pamiętaj o #dwthon #hackoutsidethebox ;) 

Śmiało nas oznaczaj, będzie nam miło :))


## 🙌 Życzymy wszystkiego dobrego, wykorzystaj moc ML do dobrych spraw, o których możesz podyskutować w tym kanale:  [the_world_needs_ml](https://bit.ly/3vgNfFa) 🙌 

Dowiesz się tam także więcej o [Fundacji DataWorkshop](https://bit.ly/3ewa29S) i o tym, jak można zaangażować się w jej działania :) 

## 🤖 Obejrzyj webinarium: [10 najpopularniejszych błędów w uczeniu maszynowym](https://bit.ly/3qKQbqg) 🤖 